File: FlutterApp/lib/domain/entities/chat_message.dart
import 'package:equatable/equatable.dart';
import '../entities/file_attachment.dart';

enum MessageType { text, file, image, video, system }
enum MessageStatus { sent, delivered, read, deleted }

class ChatMessage extends Equatable {
  final String id;
  final String senderId;
  final String senderName;
  final String? senderAvatar;
  final String? receiverId;
  final String? groupId;
  final String? textContent;
  final MessageType messageType;
  final MessageStatus status;
  final List<FileAttachment> attachments;
  final DateTime createdAt;
  final DateTime? updatedAt;
  
  const ChatMessage({
    required this.id,
    required this.senderId,
    this.senderName = '',
    this.senderAvatar,
    this.receiverId,
    this.groupId,
    this.textContent,
    this.messageType = MessageType.text,
    this.status = MessageStatus.sent,
    this.attachments = const [],
    required this.createdAt,
    this.updatedAt,
  });
  
  bool get isGroupMessage => groupId != null;
  bool get hasAttachments => attachments.isNotEmpty;
  bool get isMine => senderId == currentUserId;
  
  static String currentUserId = '';
  
  ChatMessage copyWith({
    MessageStatus? status,
    List<FileAttachment>? attachments,
    String? textContent,
  }) {
    return ChatMessage(
      id: id,
      senderId: senderId,
      senderName: senderName,
      senderAvatar: senderAvatar,
      receiverId: receiverId,
      groupId: groupId,
      textContent: textContent ?? this.textContent,
      messageType: messageType,
      status: status ?? this.status,
      attachments: attachments ?? this.attachments,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
  
  @override
  List<Object?> get props => [id, senderId, messageType, status, createdAt];
}

File: ChatWithFiles.Domain/Entities/ChatMessage.cs
using ChatWithFiles.Domain.Common;
using ChatWithFiles.Domain.Enums;

namespace ChatWithFiles.Domain.Entities;

public class ChatMessage : BaseEntity
{
    public Guid SenderId { get; set; }
    public Guid? GroupId { get; set; }
    public Guid? ReceiverId { get; set; }
    public string? TextContent { get; set; }
    public MessageType MessageType { get; set; } = MessageType.Text;
    public MessageStatus Status { get; set; } = MessageStatus.Sent;
    
    // Metadata
    public string? IpAddress { get; set; }
    public string? UserAgent { get; set; }
    
    // Navigation properties
    public User? Sender { get; set; }
    public User? Receiver { get; set; }
    public ChatGroup? Group { get; set; }
    public ICollection<FileAttachment> Attachments { get; set; } = new List<FileAttachment>();
    
    // Helper methods
    public bool IsGroupMessage() => GroupId.HasValue;
    public bool IsDirectMessage() => ReceiverId.HasValue;
    
    public void MarkAsDelivered()
    {
        Status = MessageStatus.Delivered;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public void MarkAsRead()
    {
        Status = MessageStatus.Read;
        UpdatedAt = DateTime.UtcNow;
    }
}
